pkg/dataobj/internal/dataset/page.go

```diff
@@ -0,0 +1,104 @@
+package dataset
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+)
+
+// Helper types.
+type (
+	// PageData holds the raw data for a page.
+	PageData []byte
+
+	// PageInfo describes a page.
+	PageInfo struct {
+		UncompressedSize int    // UncompressedSize is the size of a page before compression.
+		CompressedSize   int    // CompressedSize is the size of a page after compression.
+		CRC32            uint32 // CRC32 checksum of the page after encoding and compression.
+		RowCount         int    // RowCount is the number of rows in the page, including NULLs.
+
+		Value       datasetmd.ValueType       // Type of values stored in the page.
+		Compression datasetmd.CompressionType // Compression used for values in the page.
+		Encoding    datasetmd.EncodingType    // Encoding used for values in the page.
+		Stats       *datasetmd.Statistics     // Optional statistics for the page.
+	}
+)
+
+// MemPage holds an encoded (and optionally compressed) sequence of [Value]
+// entries of a common type. Use [ColumnBuilder] to construct sets of pages.
+type MemPage struct {
+	// Information about the page.
+	Info PageInfo
+
+	// Data for the page. Data is formatted as:
+	//
+	//   [uvarint(bitmapSize)][presenceBitmap][valueData]
+	//
+	// The presenceBitmap is a bitmap-encoded sequence of booleans, where values
+	// describe which rows are present (1) or nil (0). presenceBitmap is always
+	// stored uncompressed.
+	//
+	// valueData is then the encoded and optionally compressed sequence of
+	// non-NULL values, whose type, compression, and encoding are specified by Value, Compression, and Encoding.
+	Data PageData
+}
+
+var checksumTable = crc32.MakeTable(crc32.Castagnoli)
+
+// reader returns a reader for decompressed page data. Reader returns an error
+// if the CRc32 fails to validate.
+func (p *MemPage) reader() (presence io.ReadCloser, values io.ReadCloser, err error) {
```

Comment:
I suspect this will need a reset based on how many readers are used here (bytes,zstd or snappy....)

Reply:
WDYM reset? 

Good call on the presence reader, I'll change that to io.Reader.

Reply:
Going to merge since I addressed your other feedback, we can get back to the reset need in a follow-up